### PR TITLE
Notifies Jira tickets when related Eigen betas are deployed

### PR DESCRIPTION
**Pull request by ashfurrow**
As discussed back in [December's Front-End iOS Practice meeting](https://www.notion.so/artsy/December-7-2018-Meeting-Notes-a1428a9d88924551883e9c9eebfdbc97), we want to automate a notification on Jira tickets when changes related to those tickets get deployed in an Eigen beta. There was interest in doing this for Force as well, but I wanted to keep the scope limited for the initial implementation.

This is a WIP because I need to write some unit tests. I wanted to get my changes up for some early feedback. I found the [existing implementation](https://github.com/danger/peril-settings/blob/b306662628f1a4c5c4ebc5f0531e15561fe60fa5/org/new_tag.ts#L12-L64) that @orta linked to very helpful.

### CODE REVIEW

**Comment by orta** [comment]
I wonder if we can re-use a Jira field for this? (mainly thinking about the amount of notifications it could create )

**Comment by ashfurrow** [comment]
You mean like have a label or something that says "deployed" on the ticket? That's a good idea, too, though I do like the record of which beta version the fix is included in (a lot of miscommunication around what's been deployed and what's not been yet comes down to a version mismatch, and making that information available to non-iOS/non-engineers would help).

File: org/newEigenBeta.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import * as IssueJSON from "../fixtures/jira_issue_example.json"
+type Issue = typeof IssueJSON
+
+const companyPrefix = "artsyproduct"
+
+// Implementation borrowed from https://github.com/danger/peril-settings/blob/b306662628f1a4c5c4ebc5f0531e15561fe60fa5/org/new_tag.ts#L12-L64
+export default async (create: Create) => {
+  if (create.ref_type !== "tag") {
+    return console.log("Skipping because it's not a tag")
+  }
```
Comment (ashfurrow) [inline org/newEigenBeta.ts]: We should probably also skip tags that have the `-submission` suffix, since they aren't betas.

**Comment by orta** [comment]
I mean something like these fields:

screen shot 2019-01-29 at 6 18 48 pm

Which can make the data structured (or maybe both, I guess, it's not like many Jira tickets are fixed by a native release - most of that occurs in emission nowadays)

File: org/newEigenBeta.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  const numberExtractor = /Merge pull request #(\d*)/
+  const commitMessages = compareData.commits.map(c => c.commit.message)
+  const prMerges = commitMessages.filter(message => message && message.startsWith("Merge pull request #"))
+
+  // This is now a number array of PR ids
+  // e.g. [ 930, 934, 937, 932, 938 ]
+  const prs = prMerges
+    .map(msg => msg.match(numberExtractor) && msg.match(numberExtractor)![1])
+    .filter(pr => pr)
+    .map((pr: any) => parseInt(pr))
```
Comment (jonallured) [inline org/newEigenBeta.ts]: I'd like to offer a refactor here - take or leave any or all of this!! ❤️ 

```typescript
const prs = prMerges.map(message => {
  const matches = message.match(numberExtractor) || []
  const capture = matches[1]
  return parseInt(capture)
}).filter(Boolean)
```

Comment (ashfurrow) [reply org/newEigenBeta.ts]: This is great – I obviously trusted you that it works but I took a minute to break down exactly _why_ it works at each line, and it's quite elegant!

Comment (jonallured) [reply org/newEigenBeta.ts]: Yay!! I have to admit that I did not actually write that locally and test it, so I probably should have had a disclaimer to that effect! 😝 

And nice job with ac791e2 - that default export reads really nice!! 😎

**Comment by orta** [comment]
Not to keep piling on (but the ideas are pretty good) GitHub has a thing called milestones that we've never really used that we can also use this for

screen shot 2019-02-03 at 7 43 55 am

**Comment by orta** [comment]
Is this good, should we merge?

**Comment by ashfurrow** [comment]
Yeah, let's merge for now and figured out the structured field stuff later – we get 90% of the value with the implementation we currently have.

**Comment by ashfurrow** [comment]
Hmm! Seeing a [CI failure](https://travis-ci.org/artsy/peril-settings/builds/496007736) on the merge to master:

```
$ tsc  --noEmit
org/newEigenBeta.ts:4:29 - error TS7016: Could not find a declaration file for module 'semver-sort'. '/home/travis/build/artsy/peril-settings/node_modules/semver-sort/index.js' implicitly has an 'any' type.
  Try `npm install @types/semver-sort` if it exists or add a new declaration (.d.ts) file containing `declare module 'semver-sort';`
4 import * as semverSort from "semver-sort"
                              ~~~~~~~~~~~~~
org/newEigenBeta.ts:38:43 - error TS2339: Property 'getTags' does not exist on type '{ list: { (params?: ReposListParams | undefined): Promise<Response<any>>; endpoint: Endpoint; }; listForUser: { (params?: ReposListForUserParams | undefined): Promise<Response<any>>; endpoint: Endpoint; }; ... 125 more ...; deleteHook: { ...; }; }'.
38   const allTagsResponse = await api.repos.getTags(repo)
                                             ~~~~~~~
org/newEigenBeta.ts:78:36 - error TS2339: Property 'pullRequests' does not exist on type 'Octokit'.
78       const prResponse = await api.pullRequests.get({ ...repo, number: pr })
                                      ~~~~~~~~~~~~
error Command failed with exit code 1.
```

Looks like we need the semver-sort typings explicitly, but also the typings for Octokit aren't working? Curious why this didn't happen on the PR's CI run.

**Comment by ashfurrow** [comment]
Okay, I couldn't figure this out quickly. I don't have a tonne of time to work on this today – @orta  should we revert and I can come back later? LD work is a higher priority for me right now.

**Comment by orta** [comment]
Yeah, this won't work, it needed a rebase because the dependencies changed underneath it

**Comment by ashfurrow** [comment]
I've reverted this merge and will come back to this PR when I have time 👍

**Pull request merged**